### PR TITLE
GraphQL addition

### PR DESCRIPTION
**Pull request by ramth05**
**Razorpay GraphQl Extension**

Order flow for placing Magento Order using razorpay as payment method with graphql

1. set Payment Method on Cart
```
mutation {
  setPaymentMethodOnCart(input: {
      cart_id: "{{cart_ID}}"
      payment_method: {
          code: "razorpay"
      }
  }) {
    cart {
      selected_payment_method {
        code
      }
    }
  }
}
```

2. Create Razorpay Order ID against the cart 
```
mutation {
  placeRazorpayOrder (
    cart_id: "{{cartid}}"
  ){
    success
    rzp_order_id
    order_quote_id
    amount
    currency
    message
  }
}
```

3. Use `rzp_order_id` and other details from step-2 and create from the Frontend/React/using razorpay's checkout.js , complete the payment and obtain razorpay_payment_id & razorpay_signature
  https://razorpay.com/docs/payment-gateway/web-integration/standard/

4. Save Razorpay Response Details against Cart after payment success 
```
mutation {
  setRzpPaymentDetailsOnCart (
    input: {
      cart_id: "{{cart_ID}}"
      rzp_payment_id: "{{RAZORPAY_PAYMENT_ID}}"
      rzp_order_id: "{{RAZORPAY_ORDER_ID}}"
      rzp_signature: "{{RAZORPAY_SIGNATURE}}"
    }
  ){
  cart{
    id
  }
  }
}
```
5. Finally Place Magento Order 
```
mutation {
  placeOrder(input: {cart_id: "{{cart_ID}}"}) {
    order {
      order_number
    }
  }
}
```

### CODE REVIEW

**Comment by shreyas-codes** [comment]
Where are you validating razorpay signature here ?

**Comment by ramth05** [comment]
> Where are you validating razorpay signature here ?



> Where are you validating razorpay signature here ?

@shreyas-codes , As graphQL can be executed from any system, So signature validation can not be implemented here. 
We are fetching the payment again from API and doing the validation.

**Comment by ramth05** [comment]
> > Where are you validating razorpay signature here ?
> 
> > Where are you validating razorpay signature here ?
> 
> @shreyas-codes , As graphQL can be executed from any system, So signature validation can not be implemented here.
> We are fetching the payment again from API and doing the validation.

@shreyas-codes , have you tested it,

**Comment by ChetanGN** [comment]
@ramth05 order history comment not coming properly please check

File: Model/PaymentMethod.php [hunk 1]
```diff
@@ -225,21 +225,114 @@ public function authorize(InfoInterface $payment, $amount)
...
-                {
-                    $rzpOrderAmount = $order->getOrderCurrency()->formatTxt(number_format($this->order->getRazorpayOrderAmount() / 100, 2, ".", ""));
+                    $rzp_order_id = $this->order->getOrderId();
 
-                    throw new LocalizedException(__("Cart order amount = %1 doesn't match with amount paid = %2", $order->getOrderCurrency()->formatTxt($order->getGrandTotal()), $rzpOrderAmount));
-                }
+                    //validate RzpOrderamount with quote/order amount before signature
+                    $orderAmount = (int) (number_format($order->getGrandTotal() * 100, 0, ".", ""));
+
+                    if ($orderAmount !== $this->order->getRazorpayOrderAmount())
```
Comment (ChetanGN) [inline Model/PaymentMethod.php:331]: @ramth05 Can we apply the same check condition in GraphQL !

File: Model/PaymentMethod.php [hunk 2]
```diff
@@ -225,21 +225,114 @@ public function authorize(InfoInterface $payment, $amount)
...
+                    {
+                        $rzpOrderAmount = $order->getOrderCurrency()->formatTxt(number_format($this->order->getRazorpayOrderAmount() / 100, 2, ".", ""));
 
-                $this->validateSignature($request);
+                        $abcAmount = $order->getGrandTotal();
```
Comment (ChetanGN) [inline Model/PaymentMethod.php]: @ramth05 remove this variable it's not used.

**Review by ChetanGN — APPROVED** [review]
LGTM

**Pull request merged**